lib/spack/spack/solver/asp.py

```diff
@@ -115,8 +115,30 @@ def getter(node):
     "VersionProvenance", version_origin_fields
 )(**{name: i for i, name in enumerate(version_origin_fields)})
 
-#: Named tuple to contain information on declared versions
-DeclaredVersion = collections.namedtuple("DeclaredVersion", ["version", "idx", "origin"])
+
+class DeclaredVersion(object):
+    def __init__(self, version: spack.version.VersionBase, idx, origin):
+        if not isinstance(version, spack.version.VersionBase):
+            raise ValueError("Unexpected type for declared version: {0}".format(type(version)))
+        self.version = version
+        self.idx = idx
+        self.origin = origin
+
+    def __eq__(self, other):
+        if not isinstance(other, DeclaredVersion):
+            return False
+        is_git = lambda v: isinstance(v, spack.version.GitVersion)
+        if is_git(self.version) != is_git(other.version):
+            # GitVersion(hash=x) and Version(y) compare equal if x == y, but
```

Comment:
It's actually not possible to have a `Version` with the git syntax b/c there is no `Version`, only `VersionBase`, but VersionBase is not part of the public API:

```python
>>> from spack.version import ver
>>> ver("git.master-1.0")
GitVersion('git.master-1.0')
>>> spack.version.Version("git.master-1.0")
GitVersion('git.master-1.0')
```

```python
>>> from spack.version import *
>>> VersionBase("git.master=foo")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
NameError: name 'VersionBase' is not defined
```

We *do* use `VersionBase` in a few places outside `version.py`, but not to construct anything; only to test if something is a version and to copy it with `Version`.  That is likely the result of some refactoring that could be cleaned up but it doesn't affect this.

The parser also will not produce a `VersionBase` with a git version in it.

So basically this test should be unnecessary -- we don't allow it.

Reply:
"bug" implies that things aren't working as intended, but part of https://github.com/spack/spack/pull/35057 was to establish and enforce what the intent *is* (e.g. `test_git_user_supplied_reference_satisfaction` was added in https://github.com/spack/spack/pull/35057)

if you have `hash=2.2` and `2.2`, then intuitively (and enforced by the code) they are equal: we are saying:

* "I have something equivalent to X"
* "I have X"

So I think it makes sense for them to be equal. based on other conversations, the satisfaction semantics need to be different. https://github.com/spack/spack/pull/35057 ensures that all of those intentions are enforced. If you disagree with the intent then that merits a conversation about the desired semantics.

Reply:
Equating those makes things hard to reason about and will definitely lead to subtle bugs. For example:

```python
>>> spack.version.Version("1.0") in set(spack.version.GitVersion("abcdef=1.0"))
False
>>> spack.version.Version("1.0") in [spack.version.GitVersion("abcdef=1.0")]
True
```

Reply:
Note this does not have to do with the discrepancy between `==` and `satisfies`, but rather with a failure to implement consistency between `==` and `.hash()`, you can see that if you replace

```
spack.version.Version("1.0") in [spack.version.GitVersion("abcdef=1.0")]
```

with 

```
spack.version.Version("1.0") in set([spack.version.GitVersion("abcdef=1.0")])
```

which will also return `False`. These edits change nothing about `==` for versions so this bug is in fact preexisting.

Reply:
Sure, it's pre-existing, but then imo it would be better to fix the bug instead of working around it in the `exact_match` related changes in this PR.

Reply:
@haampie this concern is ostensibly resolved in https://github.com/spack/spack/commit/c92938fe6afa6725388dd3e4c4234b793602675b, but see https://github.com/spack/spack/pull/36347#issuecomment-1481973196 for a more-careful treatment of the related concerns

None of the problems addressed by this PR have to do with the construction of sets, so cannot be resolved by making `hash` and `==` consistent. I've asserted here that the semantics for `==` and `satisfies` are what they should be, and have seen no argument in this PR making a case for (or convincing me) otherwise, so have trouble understanding how "fixing" that is the solution (if it in fact would resolve this problem).

Reply:
Right, I'm saying the exact match logic could've been easier if one wouldn't have to worry about version == git version. Since it's just one instance now, and the long comment could be dropped, it's fine.